execution/executor/src/chunk_executor.rs

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::{
+    components::{
+        chunk_commit_queue::ChunkCommitQueue, chunk_output::ChunkOutput,
+        executed_chunk::ExecutedChunk,
+    },
+    logging::{LogEntry, LogSchema},
+    metrics::{
+        DIEM_EXECUTOR_APPLY_CHUNK_SECONDS, DIEM_EXECUTOR_COMMIT_CHUNK_SECONDS,
+        DIEM_EXECUTOR_EXECUTE_CHUNK_SECONDS,
+    },
+};
+use anyhow::Result;
+use diem_infallible::Mutex;
+use diem_logger::prelude::*;
+use diem_state_view::StateViewId;
+use diem_types::{
+    contract_event::ContractEvent,
+    ledger_info::LedgerInfoWithSignatures,
+    protocol_spec::DpnProto,
+    transaction::{
+        default_protocol::{TransactionListWithProof, TransactionOutputListWithProof},
+        TransactionInfo,
+    },
+};
+use diem_vm::VMExecutor;
+use executor_types::{ChunkExecutorTrait, ExecutedTrees};
+use fail::fail_point;
+use std::{marker::PhantomData, sync::Arc};
+use storage_interface::{default_protocol::DbReaderWriter, state_view::VerifiedStateView};
+
+pub struct ChunkExecutor<V> {
```

Comment:
I'm deduplicating by extracting out what's in common -- "executing transactions", "processing vm output", "applying vm output to a speculative view", "committing". And I'm separating the two executors because they manage the pending transactions in two different structures: a block tree, and a chunk queue.

Reply:
Makes sense! I do wonder if we could ever move to a world where we could eliminate/minimize the block vs. chunk difference, but this is a good improvement.

Reply:
My thinking is the queue is kept for state sync and db-restore, and after we don't have the DEC ;-), we remove the block managing ability (consensus manages it, and keeps ExecutedChunk on its block tree)